### ISSUE TITLE
Bug : It is not possible to add a third party contact to a project if the project is not linked to a third party.

### ISSUE THREAD

**Issue by bgenere** [Dolibarr/dolibarr#6184]
Actually this is a functional issue not a technical one.

Project and task contacts could be Dolibarr users  and or contact (from a third party). However I found out a code restriction that do not allow to add contact to a project or a task when the project is not linked to a third party. In my opinion this is a bug.

`// Line to add an external contact. Only if project linked to a third party.
// if ($projectstatic->socid) {`

What does mean linking a project to a third party ? I think that should be the case, when the project is planned to be done for this third party. I.e you are setting a plan for a client and the project will be done for this client. This is what I call an external project.

However you may need to follow internal projects, somes that you are doing for your company. As an example, you need to set-up Dolibarr internaly and you want to plan and track what is done.

I am currently working on a module where projects are used to monitor cultivation tasks for farmers. So here we have internal projects which are not associated to a specific third party. However, to deliver this project internally, we need to get external people i.e. contacts from suppliers working on some tasks. We want to give them a plan of what they have to do and then track the time they have spent.

So basically we should not have this restriction in code. I also notice that this restriction has no sense today has if you add a third party to your project , then you could add any contact of any third party to the project. So a current work-around is to create a generic "internal" third party to be able to manage internal project. However this is not really elegant, add dummy data.

So i am in favour of removing this useless code.
 
(I already made a copy of this code with the control removed for my custom module as I need a simple interface, and it works perfectly).

Let me know if you agree. I will be happy to make the change and summit my work.

**Comment by bgenere** [comment]
Actually, I am going deeper in the module and I get confused.

You could assign a contact to a project or a task, but it is not possible to track time for this contact. The table projet_task_time contains only  the fk_user reference but has no field to store fk_contact.

So the only way to store directly time task for a contact, is to create a user for it, meaning that selecting a contact for a task or a project does not make any sense. 

So I see 2 options currently:

1. We keep the ability to associate user and contact to project/task and we extend the project_task_time to add an fk_contact field and store time spend by contact.

2. We  remove the ability to associate a contact to project/task and we document the fact that contact working on project should have a Dolibarr user created. So they could be selected on project/task and we could track their time.

Option 2 will simplify the interface and the code, however it implies more work to create and administrate the user accounts for the administrator. The change could also be seen as a regression even it is actually not as associating directly a contact to a project/task does not make much sense today.

Option 1 extend the module feature and complete the work done on adding contact to project/task.

Let me know what do you think is the best option.

**Comment by eldy** [comment]
"1.We keep the ability to associate user and contact to project/task and we extend the project_task_time to add an fk_contact field and store time spend by contact." => mixing apple and pear does not seems to be a good idea. The 2 is clearly what we want in Dolibarr:
If we want to track time, it means we must have a user.

Adding a contact that is not a user is a feature available for information only. Not for tracking time. If you need to track time, yes, you must create a user. Note that creating a user from a contact is already possible.

**Comment by bgenere** [comment]
I also agree that solution 1 is actually complex and confusing. I found that creating a user from a contact is easy and solve the issue.

When version 5 will be out I will document the project module with this rule. I will then look at the code and will propose  some changes. 

I am currently working on a simplified project interface for the [vignoble module](https://github.com/bgenere/vignoble). When complete, may be some of the code done there could be included as default in the project core module.

**Comment by bgenere** [comment]
I close this issue as this is now documented on the wiki that only user could track their time and contact is just for documentation.
Also closed because bug #6241 is more important and urgent in my opinion.
